### PR TITLE
Make ProductAttributesContainer a lazy object so that initialisation is simpler.

### LINKED ISSUE #3258: Writing to products attributes with Product.attr can lead to dataloss
### Issue Summary

Writing to products attributes with `product.attr` can lead to data loss

### Steps to Reproduce

Create a productclass with 3 attributes, the second one you create must be optional.
Create a product and do not fill the second optional attribute. Now write to the non-optional attributes with code like this:

        p = Product.objects.get(pk=4)
        p.attr.subtitle = "lama"
        p.attr.available = False
        p.attr.save()

One of the 2 attributes will not be updated because the attr object will not be intialized. During save it will encounter the second attribute which is not on the object and trigger initialization. Initialization will overwrite one of the 2 values written and reset it to whatever is in the database.

### Technical details

This problem occurs with all python versions and all oscar versions.

### PR DESCRIPTION
**Pull request by solarissmoke**
Fixes #3258.

### CODE REVIEW

File: src/oscar/apps/catalogue/product_attributes.py
Comment (solarissmoke) [inline src/oscar/apps/catalogue/product_attributes.py:24]: You get an infinite loop if this is set *after* the loop below, because that itself is using `setattr`...

**Review by auvipy — APPROVED** [review]

**Comment by solarissmoke** [comment]
@auvipy request that you please refrain from approving PRs unnecessarily - if you have feedback on the changes then please feel free to comment, otherwise there's no need to leave a review (it just creates noise for people who receive notifications), as we still need someone from the core team to approve. Thanks.

**Review by specialunderwear — APPROVED** [review]
Nice work!

**Pull request merged**
